aws/resource_aws_gamelift_fleet.go

```diff
@@ -0,0 +1,596 @@
+		d.SetId("")
+		return nil
+	}
+	if len(attributes) != 1 {
+		return fmt.Errorf("Expected exactly 1 Gamelift fleet, found %d under %q",
+			len(attributes), d.Id())
+	}
+	fleet := attributes[0]
+
+	d.Set("build_id", fleet.BuildId)
+	d.Set("description", fleet.Description)
+	d.Set("fleet_arn", fleet.FleetArn)
+	d.Set("metric_groups", flattenStringList(fleet.MetricGroups))
+	d.Set("name", fleet.Name)
+	d.Set("new_game_session_protection_policy", fleet.NewGameSessionProtectionPolicy)
+	d.Set("operating_system", fleet.OperatingSystem)
+	d.Set("resource_creation_limit_policy", flattenGameliftResourceCreationLimitPolicy(fleet.ResourceCreationLimitPolicy))
+
+	return nil
+}
+
+func resourceAwsGameliftFleetUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).gameliftconn
+
+	log.Printf("[INFO] Updating Gamelift Fleet: %s", d.Id())
+
+	if d.HasChange("description") || d.HasChange("metric_groups") || d.HasChange("name") ||
+		d.HasChange("new_game_session_protection_policy") || d.HasChange("resource_creation_limit_policy") {
+		_, err := conn.UpdateFleetAttributes(&gamelift.UpdateFleetAttributesInput{
+			Description:  aws.String(d.Get("description").(string)),
+			FleetId:      aws.String(d.Id()),
+			MetricGroups: expandStringList(d.Get("metric_groups").([]interface{})),
+			Name:         aws.String(d.Get("name").(string)),
+			NewGameSessionProtectionPolicy: aws.String(d.Get("new_game_session_protection_policy").(string)),
+			ResourceCreationLimitPolicy:    expandGameliftResourceCreationLimitPolicy(d.Get("resource_creation_limit_policy").([]interface{})),
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	if d.HasChange("ec2_inbound_permission") {
+		oldPerms, newPerms := d.GetChange("ec2_inbound_permission")
+		authorizations, revocations := diffGameliftPortSettings(oldPerms.([]interface{}), newPerms.([]interface{}))
+
+		_, err := conn.UpdateFleetPortSettings(&gamelift.UpdateFleetPortSettingsInput{
+			FleetId: aws.String(d.Id()),
+			InboundPermissionAuthorizations: authorizations,
+			InboundPermissionRevocations:    revocations,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	if d.HasChange("runtime_configuration") {
+		_, err := conn.UpdateRuntimeConfiguration(&gamelift.UpdateRuntimeConfigurationInput{
+			FleetId:              aws.String(d.Id()),
+			RuntimeConfiguration: expandGameliftRuntimeConfiguration(d.Get("runtime_configuration").([]interface{})),
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return resourceAwsGameliftFleetRead(d, meta)
+}
+
+func resourceAwsGameliftFleetDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).gameliftconn
+
+	log.Printf("[INFO] Deleting Gamelift Fleet: %s", d.Id())
+	// It can take ~ 1 hr as Gamelift will keep retrying on errors like
+	// invalid launch path and remain in state when it can't be deleted :/
+	err := resource.Retry(60*time.Minute, func() *resource.RetryError {
```

Comment:
Not sure if I have a good answer here, but it would be nice to tie this to the schema timeout because up to an extra hour of apply is an awful long time 🙁

Reply:
Generally speaking we should only let users customize the timeout when we know that _intentionally_ different config (like bigger instance) requires higher timeout. This typically _doesn't_ involve error retries, because these are not affected by instance sizes or anything like that, these are always going to be the same for every user.

I see what you're saying here though and I have mixed feelings about that, but is returning error early any better? Basically you would _already know_ that the fleet has errors which would be raised during creation (that has sufficient timeout). Here the user is _knowingly_ deleting a fleet that is already in such broken state, so I'd argue that waiting is desired? 🤔

Reply:
Your argument is perfectly right - I think where I'm coming from would be fixed more with the Terraform core enhancement to be able to provide apply-time status messaging during (long) running operations.